### DISCUSSION TITLE
Version-free download link for latest EMF snapshot

### DISCUSSION THREAD

**Discussion by trancexpress** [eclipse-emf/org.eclipse.emf#33] [General]
I'd like to be able to download the latest EMF snapshot without referring to any version.

E.g. there is this update site:

https://download.eclipse.org/modeling/emf/emf/builds/nightly/latest/

I can use e.g. `p2` to mirror it locally, which is great, but requires an Eclipse installation.

Ideally there would be a redirect to the latest build, similar to the download button of the update site webpage. Currently, this download button leads to:

https://download.eclipse.org/modeling/emf/emf/builds/nightly/N202403141008/EMF-Updates-N202403141008.zip

It would be great to have something like this:

https://download.eclipse.org/modeling/emf/emf/builds/nightly/latest-snapshot/EMF-Updates-latest-snapshot.zip

**Comment by trancexpress** [comment]
Hi @merks ,

is there already a download link without a version string in it, for the latest EMF snapshot?

The background:

We do a latest integration build testing for Eclipse+xtext, one of the bits we need but don't build ourselves is EMF. We do download the snapshot (with plain `wget`) and distribute it internally.

Right now I'm reading the update site `index.html` to get the download link for the latest snapshot :( Since I can't rely on an Eclipse installation at a known location, to use `p2`. It would be great to just have one link for the snapshot, similar to how there is one update site for the latest EMF.

Best regards and thanks,
Simeon

**Reply by merks** [reply]
I'm not sure if there is a way to "implement" this without generating an additional update site zip with an fixed/known location.  I'd like to avoid  to generate such additional footprint.

I think something like this redirection approach only works in browsers but not with wget though I'm not sure:

https://stackoverflow.com/questions/5411538/how-to-redirect-one-html-page-to-another-on-load

PHP would work, but I don't want to introduce PHP where non exists already.

Note that the Eclipse Installer also contains the p2 mirror application:

https://download.eclipse.org/justj/?file=oomph/products

It exists at a fixed URI.

You could use it like this:
```
#!/bin/sh

wget -q -O application.zip https://download.eclipse.org/oomph/products/eclipse-inst-jre-win64.zip
unzip application.zip -d application

eclipse="application/eclipse-inst.exe"

if [[ $# != 2 ]]; then
  echo "mirror-eclipse <source-URL> <targetURL>"
else
 $eclipse -nosplash -verbose \
   -application org.eclipse.equinox.p2.artifact.repository.mirrorApplication \
   -source $1 \
   -destination $2
 $eclipse -nosplash -verbose \
   -application org.eclipse.equinox.p2.metadata.repository.mirrorApplication \
   -source $1 \
   -destination $2
fi

exit
```

That solution works for any update site.

Another fixed location is this one:

https://download.eclipse.org/modeling/emf/emf/builds/nightly/latest/compositeContent.jar

From that you can get the version number more directly and in a way that uses a known/fixed p2 representation:
```
$unzip -p compositeContent.jar compositeContent.xml | grep -Po "<child location='../(.*)'"
<child location='../N202403141008'
```

That version you can use to create the UR:
```
https://www.eclipse.org/downloads/download.php?file=/modeling/emf/emf/builds/nightly/${0}/EMF-Updates-${0}.zip
```

If someone wants something different, open to suggestions, but, as I said, I prefer not to maintain copies of files.

**Reply by trancexpress** [reply]
Alright, thank you. I'll leave it at HTML parsing, then eventually switch to parsing the `compositeContent.xml` as you suggest. Thanks!

**Reply by merks** [reply]
FYI, I'm asking the Foundation IT folks if there is some way to express a redirection that would work with wget/curl...
